File: Algorithms/115_DistinctSubsequences.py
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        n, m = len(s), len(t)
        result = [[0] * (n+1) for i in range(m+1)]
        result[0][0] = 1
        for i in range(1, m+1):
            result[i][0] = 0
        for j in range(1, n+1):
            result[0][j] = 1
        for i in range(1, m+1):
            for j in range(1, n+1):
                result[i][j] = result[i][j-1]
                if s[j-1] == t[i-1]:
                    result[i][j] += result[i-1][j-1]
        return result[m][n]

File: Algorithms/115.DistinctSubsequences.cpp
class Solution {
public:
    int numDistinct(string S, string T) {
        int n = S.size();
        int m = T.size();
        int result[m+1][n+1];
        result[0][0] = 1;
        for (int i=1; i <= m; ++i) {
            result[i][0] = 0;
        }
        for (int j=1; j <= n; ++j) {
            result[0][j] = 1;
        }
        for (int i=1; i <= m; ++i) {
            for (int j=1; j <= n; ++j) {
                result[i][j] = result[i][j-1];
                if (S[j-1] == T[i-1]) {
                    result[i][j] += result[i-1][j-1];
                }
            }
        }
        return result[m][n];
    }
};
